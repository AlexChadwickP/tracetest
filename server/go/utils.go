package openapi

import (
	"net/http"

	"github.com/gogo/protobuf/jsonpb"
	"github.com/gogo/protobuf/proto"
	v11 "go.opentelemetry.io/proto/otlp/common/v1"
	res "go.opentelemetry.io/proto/otlp/resource/v1"
	v1 "go.opentelemetry.io/proto/otlp/trace/v1"
)

func EncodeJSONPBResponse(i interface{}, status *int, w http.ResponseWriter) error {
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	if status != nil {
		w.WriteHeader(*status)
	} else {
		w.WriteHeader(http.StatusOK)
	}

	trace := i.(proto.Message)
	m := jsonpb.Marshaler{}
	return m.Marshal(w, trace)
}

func FixParent(tr *v1.TracesData, traceID, parentSpanID string) *v1.TracesData {
	spans := make(map[string]*v1.Span)
	for _, rs := range tr.ResourceSpans {
		for _, ils := range rs.InstrumentationLibrarySpans {
			for _, sp := range ils.Spans {
				spans[string(sp.SpanId)] = sp
			}
		}
	}

	// Fix parent id
	for _, sp := range spans {
		if sp.ParentSpanId == nil {
			continue
		}
		_, ok := spans[string(sp.ParentSpanId)]
		if !ok {
			sp.ParentSpanId = []byte(parentSpanID)
		}
	}
	rs := &v1.ResourceSpans{
		Resource: &res.Resource{
			Attributes:             []*v11.KeyValue{},
			DroppedAttributesCount: 0,
		},
		InstrumentationLibrarySpans: []*v1.InstrumentationLibrarySpans{
			{
				InstrumentationLibrary: &v11.InstrumentationLibrary{
					Name:    "tracetest",
					Version: "v1",
				},
				Spans: []*v1.Span{
					{TraceId: []byte(traceID),
						SpanId:       []byte(parentSpanID),
						ParentSpanId: nil,
						Name:         "tracetest",
						Kind:         v1.Span_SPAN_KIND_CLIENT,
					},
				},
			},
		},
		SchemaUrl: "",
	}
	tr.ResourceSpans = append(tr.ResourceSpans, rs)

	return tr
}
