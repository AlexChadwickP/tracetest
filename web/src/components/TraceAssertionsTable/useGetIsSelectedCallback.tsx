import {useCallback} from 'react';
import {useStore} from 'react-redux';

export function useGetIsSelectedCallback() {
  const store = useStore();
  return useCallback(
    (spanId: string): boolean => {
      const state = store.getState();
      const {selectedElements} = state;
      const found = selectedElements ? selectedElements.find(({id}: {id: string}) => id === spanId) : undefined;
      return Boolean(found);
    },
    [store]
  );
}
