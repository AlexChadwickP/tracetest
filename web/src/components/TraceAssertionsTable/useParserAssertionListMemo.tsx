import {difference, sortBy} from 'lodash';
import {useMemo} from 'react';
import {SpanAssertionResult} from '../../types/Assertion.types';

export interface TParsedAssertion {
  key: string;
  spanLabels: string[];
  property: string;
  comparison: string;
  value: string;
  actualValue: string;
  hasPassed: boolean;
  spanId: string;
}

export function useParserAssertionListMemo(
  spanListAssertionResult: SpanAssertionResult[],
  selectorValueList: string[]
): TParsedAssertion[] {
  return useMemo(() => {
    const spanAssertionList = spanListAssertionResult.reduce<Array<TParsedAssertion>>((list, {resultList, span}) => {
      const subResultList = resultList.map<TParsedAssertion>(
        ({propertyName, comparisonValue, operator, actualValue, hasPassed, spanId}) => {
          const spanLabelList = span.signature.map(({value}) => value).concat([`#${spanId.slice(-4)}`]) || [];

          return {
            spanLabels: difference(spanLabelList, selectorValueList),
            spanId,
            key: `${propertyName}-${spanId}`,
            property: propertyName,
            comparison: operator,
            value: comparisonValue,
            actualValue,
            hasPassed,
          };
        }
      );

      return list.concat(subResultList);
    }, []);

    return sortBy(spanAssertionList, ({spanLabels}) => spanLabels.join(''));
  }, [selectorValueList, spanListAssertionResult]);
}
