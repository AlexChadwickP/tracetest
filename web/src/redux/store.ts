import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import {testAPI} from '../services/TestService';

export const store = configureStore({
  reducer: {
    [testAPI.reducerPath]: testAPI.reducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
