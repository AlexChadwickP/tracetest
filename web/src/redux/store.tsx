import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit';
import TestAPI from 'redux/apis/TraceTest.api';

export const store = configureStore({
  reducer: {
    [TestAPI.reducerPath]: TestAPI.reducer,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TestAPI.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
