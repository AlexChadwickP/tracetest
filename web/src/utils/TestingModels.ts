export const TestingModels = {
  assertionResult: {
    assertion: {assertionId: '', selectors: [], spanAssertions: []},
    spanListAssertionResult: [],
  },
};
