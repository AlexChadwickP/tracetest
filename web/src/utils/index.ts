const flattenAttributesSelector = () =>
  `resourceSpans[].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0]},resource.attributes[].{key:key,value: value.*|[0]}]|[][]`;

export const filterBySpanId = (spanId: string = '') =>
  `resourceSpans[?instrumentationLibrarySpans[?spans[?starts_with(spanId,'${spanId}')]]] | [].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0],type:'span'},resource.attributes[].{key:key,value: value.*|[0],type:'resource'}]|[][]`;
