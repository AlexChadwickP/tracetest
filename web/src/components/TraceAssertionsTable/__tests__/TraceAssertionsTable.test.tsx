import {fireEvent, render} from '@testing-library/react';
import {ReactFlowProvider} from 'react-flow-renderer';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import {TestingModels, TestingObj} from '../../../utils/TestingModels';
import TraceAssertionsTable from '../TraceAssertionsTable';

test('TraceAssertionsTable', () => {
  const result = render(
    <ReactFlowProvider>
      <ReduxWrapperProvider>
        <TraceAssertionsTable assertionResult={TestingObj.assertionResult} onSpanSelected={jest.fn()} />
      </ReduxWrapperProvider>
    </ReactFlowProvider>
  );
  expect(result.container).toMatchSnapshot();
});

test('TraceAssertionsTable click row', () => {
  const result = render(
    <ReactFlowProvider>
      <ReduxWrapperProvider>
        <TraceAssertionsTable assertionResult={TestingObj.assertionResult} onSpanSelected={jest.fn()} />
      </ReduxWrapperProvider>
    </ReactFlowProvider>
  );

  const elementsByClassName = result.container.getElementsByClassName('ant-table-row');
  expect(result.container).toMatchSnapshot();
  fireEvent(elementsByClassName.item(0)!, TestingModels.mouseEvent);
});
