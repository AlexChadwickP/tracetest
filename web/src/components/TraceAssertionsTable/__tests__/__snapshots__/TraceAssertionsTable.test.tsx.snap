// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceAssertionsTable 1`] = `
<div>
  <div
    class="sc-gsDKAQ jfpIVA"
    data-cy="test-results-assertion-table"
  >
    <div
      class="sc-dkPtRN diGMvB"
    >
      <h5
        class="ant-typography"
        style="margin: 0px;"
      >
        
      </h5>
      <h5
        class="ant-typography"
        style="margin: 0px;"
      >
        1 span
      </h5>
    </div>
    <div
      class="ant-table-wrapper sc-bdvvtL kLJHSf"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-small ant-table-bordered ant-table-layout-fixed"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <colgroup>
                    <col
                      style="width: 40%;"
                    />
                    <col
                      style="width: 25%;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title="Span Labels"
                      >
                        Span Labels
                      </th>
                      <th
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title="Property"
                      >
                        Property
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Comparison
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Value
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Actual
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="-"
                    >
                      <td
                        class="ant-table-cell ant-table-cell-ellipsis"
                      >
                        <span
                          class="ant-badge ant-badge-not-a-wrapper sc-hKwDye wxfDf"
                        >
                          <sup
                            class="ant-scroll-number ant-badge-count"
                            data-show="true"
                            title="#"
                          >
                            #
                          </sup>
                        </span>
                      </td>
                      <td
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title=""
                      >
                        
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        eq
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        <span
                          class="ant-typography ant-typography-danger"
                        >
                          <strong>
                            New request
                          </strong>
                        </span>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TraceAssertionsTable click row 1`] = `
<div>
  <div
    class="sc-gsDKAQ jfpIVA"
    data-cy="test-results-assertion-table"
  >
    <div
      class="sc-dkPtRN diGMvB"
    >
      <h5
        class="ant-typography"
        style="margin: 0px;"
      >
        
      </h5>
      <h5
        class="ant-typography"
        style="margin: 0px;"
      >
        1 span
      </h5>
    </div>
    <div
      class="ant-table-wrapper sc-bdvvtL kLJHSf"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-small ant-table-bordered ant-table-layout-fixed"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <colgroup>
                    <col
                      style="width: 40%;"
                    />
                    <col
                      style="width: 25%;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title="Span Labels"
                      >
                        Span Labels
                      </th>
                      <th
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title="Property"
                      >
                        Property
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Comparison
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Value
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Actual
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="-"
                    >
                      <td
                        class="ant-table-cell ant-table-cell-ellipsis"
                      >
                        <span
                          class="ant-badge ant-badge-not-a-wrapper sc-hKwDye wxfDf"
                        >
                          <sup
                            class="ant-scroll-number ant-badge-count"
                            data-show="true"
                            title="#"
                          >
                            #
                          </sup>
                        </span>
                      </td>
                      <td
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title=""
                      >
                        
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        eq
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        <span
                          class="ant-typography ant-typography-danger"
                        >
                          <strong>
                            New request
                          </strong>
                        </span>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
